polyply/tests/test_lib_files.py

```diff
@@ -0,0 +1,244 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Runs more elaborate integration tests
+"""
+
+from collections import defaultdict
+from pathlib import Path
+import shlex
+import subprocess
+import sys
+import numpy as np
+
+import pytest
+import vermouth
+from vermouth.forcefield import ForceField
+
+from polyply import TEST_DATA
+from vermouth.tests.helper_functions import find_in_path
+
+INTEGRATION_DATA = Path(TEST_DATA + '/library_tests')
+
+PATTERN = '{path}/{library}/{polymer}/polyply'
+
+POLYPLY = find_in_path(names=("polyply", ))
+
+def assert_equal_blocks(block1, block2):
+    """
+    Asserts that two blocks are equal to gain the pytest rich comparisons,
+    which is lost when doing `assert block1 == block2`
+    """
+    assert block1.name == block2.name
+    assert block1.nrexcl == block2.nrexcl
+    assert block1.force_field == block2.force_field  # Set to be equal
+    # Assert the order to be equal as well...
+    # assert list(block1.nodes) == list(block2.nodes)
+    # ... as the attributes
+
+    for node in block1.nodes:
+        for attr in ["atype", "charge"]:
+            assert block1.nodes[node][attr] == block2.nodes[node][attr]
+
+    edges1 = {frozenset(e[:2]): e[2] for e in block1.edges(data=True)}
+    edges2 = {frozenset(e[:2]): e[2] for e in block2.edges(data=True)}
+
+    for e, attrs in edges2.items():
+        for k, v in attrs.items():
+            if isinstance(v, float):
+                attrs[k] = pytest.approx(v, abs=1e-3) # PDB precision is 1e-3
+
+    assert edges1 == edges2
+
+    for inter_type in ["bonds", "angles", "constraints", "exclusions", "pairs", "dihedrals", "impropers"]:
+        interactions = block1.interactions.get(inter_type, [])
+        new_interactions = block2.interactions.get(inter_type, [])
+        new_terms = defaultdict(list)
+        ref_terms = defaultdict(list)
+        assert len(interactions) == len(new_interactions)
+        for inter in interactions:
+            atoms = inter.atoms
+            ref_terms[frozenset(atoms)].append(inter)
+            for inter_new in new_interactions:
+                if set(atoms) == set(inter_new.atoms):
+                    new_terms[frozenset(atoms)].append(inter_new)
+
+        for atoms, ref_interactions in ref_terms.items():
+            new_interactions = new_terms[atoms]
+            for inter in ref_interactions:
+                print("--------------------------------------")
+                print(inter)
+                print(new_interactions)
+                print("--------------------------------------")
+                assert sum([_interaction_equal(inter, inter_new, inter_type) for inter_new in new_interactions[:len(ref_interactions)]]) == 1
```

Comment:
sum?
Plus, please break this up in more digestible chunks

Reply:
I de-convoluted this a bit, which also meant dealing with the previous comment. All these comparisons are needed because dihedrals for examples can stack and there can be at most 1 dihedral pair matching.